### ISSUE TITLE
JIT fails on creating bfloat16 fusions

### ISSUE THREAD

**Issue by slayton58** [pytorch/pytorch#53918]
## 🐛 Bug

Trying to created basic fused bf16 kernels fails.

## To Reproduce

test.py
```
import torch
torch._C._jit_set_profiling_executor(False)
torch._C._jit_set_profiling_mode(False)
def foo(x, y):
    return (x + y).relu()
m = torch.jit.script(foo)
x = torch.randn(65536).cuda().bfloat16()
y = torch.randn_like(x)
print(m.graph_for(x,y))
```

Running:
```
python test.py
```
Currently outputs (cut for brevity):
```
...
default_program(77): error: name followed by "::" must be a class or namespace name
default_program(77): error: expected a ";"
default_program(78): error: name followed by "::" must be a class or namespace name
default_program(78): error: expected a ";"
default_program(79): error: identifier "t2_buf" is undefined
default_program(79): error: identifier "n4" is undefined
default_program(76): warning: variable "n2" was declared but never referenced
...
```
## Expected behavior

Fused operator is created and appropriate graph is printed. Code does not crash.

## Environment

```
PyTorch version: 1.8.0
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 11.1.105
GPU models and configuration: GPU 0: GeForce RTX 3090
Nvidia driver version: 460.35
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.8.0
[pip3] torchelastic==0.2.2
[pip3] torchtext==0.9.0
[pip3] torchvision==0.9.0
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               11.1.74              h6bb024c_0    nvidia
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2020.2                      256
[conda] mkl-service               2.3.0            py38he904b0f_0
[conda] mkl_fft                   1.3.0            py38h54f3939_0
[conda] mkl_random                1.1.1            py38h0573a6f_0
[conda] numpy                     1.19.2           py38h54aff64_0
[conda] numpy-base                1.19.2           py38hfa32c7d_0
[conda] pytorch                   1.8.0           py3.8_cuda11.1_cudnn8.0.5_0    pytorch
[conda] torchelastic              0.2.2                    pypi_0    pypi
[conda] torchtext                 0.9.0                      py38    pytorch
[conda] torchvision               0.9.0                py38_cu111    pytorch
```

cc @gmagogsfm

**Comment by slayton58** [comment]
Note: Working with @ngimel and disabling the legacy fuser my test does not crash, but it also doesn't generate fusions for bf16 (while it does in fp32 and fp16): 
```
graph(%x.1 : Tensor,
      %y.1 : Tensor):
  %2 : int = prim::Constant[value=1]()
  %5 : Tensor = aten::add(%x.1, %y.1, %2) # test.py:5:12
  %7 : Tensor = aten::relu(%5) # test.py:5:12
  return (%7)
```

**Comment by bertmaher** [comment]
It looks like the fuser doesn't have the proper special-casing for bfloat16.  E.g., it just stringifies the type as `at::BFloat16` instead of `__nv_bfloat16`.

E.g., when I run this test with `PYTORCH_FUSION_DEBUG=1`, I see the following fused kernel signature:
```
extern "C" __global__
void kernel_0(IndexType totalElements, const TensorInfo<at::BFloat16,1> t0, const TensorInfo<at::BFloat16,1> t1, const TensorInfo<at::BFloat16,1> t2 ) {
```

**Comment by ngimel** [comment]
Yeah, the error above is for legacy fuser, and that's fine. But the real problem is that neither nnc nor nvidia fuser create bfloat16 fusions.

**Comment by bertmaher** [comment]
I should be able to add bfloat to both legacy and nnc without too much trouble (famous last words :-p ).  I assume we want to actually do computations in bfloat16, as opposed to using the legacy fuser's fp16 technique of upconverting to fp32 for the math and then downconverting to store?

**Comment by ngimel** [comment]
we should do exact same thing for bfloat16 as we do for half, upconvert to fp32 for the math and downconvert to store. Fuser is bw-limited, so whatever limited speed-up can be achieved by using bf16 (on ampere only) is not worth lost accuracy (and that's what pytorch core does anyway).

**Comment by bertmaher** [comment]
Ok, cool, that should be easy to achieve.

**Comment by bertmaher** [comment]
The legacy fuser should be fixed by the above PR, but I'm going to keep this open until I fix bfloat16 in NNC as well (hopefully early next week).

**Comment by ZolotukhinM** [comment]
This is fixed in #64196.
